File: 2345-finding-the-number-of-visible-mountains/2345-finding-the-number-of-visible-mountains.cpp
// @Author: KING-SEN
/*
One can convert peak (x, y) of a mountain to an interval representing mountain base: [x - y; x + y]. Then the problem becomes number of intervals not contained in other intervals.

*/
class Solution {
public:
    int visibleMountains(vector<vector<int>>& peaks) {
        int len = peaks.size();
        for (int i = 0; i < len; i++) {
            int x = peaks[i][0], y = peaks[i][1];
            peaks[i][0] = x - y;
            peaks[i][1] = x + y;
        }
        // Sorting is done by left boundary first is ascending order,
	    // then by right in descending order.
        sort(peaks.begin(), peaks.end(), [](auto& a, auto& b) {
            if (a[0] == b[0]) {
                return a[1] > b[1];
            }
            return a[0] < b[0];
        });
        // Count number of intervals that are entirely inside other
        int invisible_count = 0, end = 0;
        for (int i = 0; i < len - 1;) {
            end = peaks[i][1];
            int j = i + 1;
            if (peaks[i][0] == peaks[j][0] && peaks[i][1] == peaks[j][1]) {
                invisible_count++;
            }
            while (j < len) {
                int y = peaks[j][1];
                if (y > end) {
                    break;
                }
                invisible_count++;
                j++;
            }
            i = j;
        }
        return len - invisible_count;
    }
};

File: 2345-finding-the-number-of-visible-mountains/2345-finding-the-number-of-visible-mountains.py
# KING-SEN

class Solution:
    def visibleMountains(self, peaks: List[List[int]]) -> int:
        dist = [[x - y, x + y] for x, y in peaks]
        dist.sort(key=lambda d:(d[0], -d[1]))
        i, invisible = 0, 0
        #  Count number of intervals that are entirely inside other
        while i < len(dist):
            end = dist[i][1]
            j = i + 1
            if j < len(dist) and dist[i] == dist[j]:
                invisible += 1
            while j < len(dist):
                y = dist[j][1]
                if y > end:
                    break
                invisible += 1
                j += 1
            i = j
        return len(dist) - invisible